### PR TITLE
Fix bugs for ResourceDistribution

### PR DESCRIPTION
**Pull request by veophi**
Signed-off-by: veophi <vec.g.sun@gmail.com>

### Ⅰ. Describe what this PR does
1. remove forbidden namespaces "kube-system"/ "kube-public";
2. ignore "already exists" error when dry-run at webhook;
3. ignore the errors caused by "namespace-terminating";
4. change the logic when judge whether resource needs to be updated;
5. watch resource changes;
5. fix typos and etc.

### CODE REVIEW

File: pkg/controller/resourcedistribution/resourcedistribution_controller.go [hunk 1]
```diff
@@ -186,13 +186,15 @@ func (r *ReconcileResourceDistribution) distributeResource(distributor *appsv1al
...
 	resourceKind := resource.GetObjectKind().GroupVersionKind().Kind
 	newResourceHashCode := hashResource(distributor.Spec.Resource)
-
 	return syncItSlowly(matchedNamespaces, 1, func(namespace string) *UnexpectedError {
 		// 1. try to fetch existing old resource
 		oldResource := &unstructured.Unstructured{}
 		oldResource.SetGroupVersionKind(resource.GetObjectKind().GroupVersionKind())
 		getErr := r.Client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: resourceName}, oldResource)
 		if getErr != nil && !errors.IsNotFound(getErr) {
+			if errors.HasStatusCause(getErr, corev1.NamespaceTerminatingCause) {
```
Comment (furykerry) [inline pkg/controller/resourcedistribution/resourcedistribution_controller.go:195]: is it simpler to check the deletiontimestamp of the namespace?

File: pkg/webhook/resourcedistribution/validating/utils.go [hunk 2]
```diff
@@ -44,8 +42,8 @@ var (
 	// reused by controller
 	/* ADD NEW FORBIDDEN NAMESPACE HERE*/
 	ForbiddenNamespaces = []string{
-		"kube-system",
-		"kube-public",
+		//"kube-system",
```
Comment (furykerry) [inline pkg/webhook/resourcedistribution/validating/utils.go:45]: plz leave comments about why we have empty ForbiddenNamespaces

File: pkg/controller/resourcedistribution/resourcedistribution_controller.go [hunk 3]
```diff
@@ -259,7 +267,7 @@ func (r *ReconcileResourceDistribution) cleanResource(distributor *appsv1alpha1.
 		oldResource := &unstructured.Unstructured{}
 		oldResource.SetGroupVersionKind(resource.GetObjectKind().GroupVersionKind())
 		if getErr := r.Client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: resourceName}, oldResource); getErr != nil {
-			if errors.IsNotFound(getErr) {
+			if errors.IsNotFound(getErr) || errors.HasStatusCause(getErr, corev1.NamespaceTerminatingCause) {
```
Comment (furykerry) [inline pkg/controller/resourcedistribution/resourcedistribution_controller.go:270]: is it simpler to check the deletiontimestamp of the namespace?

File: pkg/controller/resourcedistribution/utils.go [hunk 4]
```diff
@@ -302,3 +297,34 @@ func listNamespacesForDistributor(handlerClient client.Client, targets *appsv1al
 
 	return matchedSet.List(), unmatchedSet.List(), nil
 }
+
+func needToUpdate(old, new *unstructured.Unstructured) bool {
+	if !reflect.DeepEqual(old.GetAnnotations(), new.GetAnnotations()) {
```
Comment (furykerry) [inline pkg/controller/resourcedistribution/utils.go:302]: what if other components update the secrets & configmap with additional annotations& labels? can we simply check other fields

**Review by furykerry — COMMENTED** [review]
/lgtm

**Review by FillZpp — COMMENTED** [review]
/approve

**Pull request merged**